causalpy/expt_diff_in_diff.py

```
@@ -0,0 +1,203 @@
+            # drop the outcome variable
+            .drop(self.outcome_variable_name, axis=1)
+            # We may have multiple units per time point, we only want one time point
+            .groupby(self.time_variable_name)
+            .first()
+            .reset_index()
+        )
+        assert not self.x_pred_control.empty
+        (new_x,) = build_design_matrices([self._x_design_info], self.x_pred_control)
+        self.y_pred_control = self.model.predict(np.asarray(new_x))
+
+        # predicted outcome for treatment group
+        self.x_pred_treatment = (
+            self.data
+            # just the treated group
+            .query(f"{self.group_variable_name} == 1")
+            # drop the outcome variable
+            .drop(self.outcome_variable_name, axis=1)
+            # We may have multiple units per time point, we only want one time point
+            .groupby(self.time_variable_name)
+            .first()
+            .reset_index()
+        )
+        assert not self.x_pred_treatment.empty
+        (new_x,) = build_design_matrices([self._x_design_info], self.x_pred_treatment)
+        self.y_pred_treatment = self.model.predict(np.asarray(new_x))
+
+        # predicted outcome for counterfactual. This is given by removing the influence
+        # of the interaction term between the group and the post_treatment variable
+        self.x_pred_counterfactual = (
+            self.data
+            # just the treated group
+            .query(f"{self.group_variable_name} == 1")
+            # just the treatment period(s)
+            .query("post_treatment == True")
+            # drop the outcome variable
+            .drop(self.outcome_variable_name, axis=1)
+            # We may have multiple units per time point, we only want one time point
+            .groupby(self.time_variable_name)
+            .first()
+            .reset_index()
+        )
+        assert not self.x_pred_counterfactual.empty
```

As in the pymc-marketing repo, It is better to raise an error than having an assert statement (jus a suggestion)

---

Done. I've removed all asserts outside of tests and replaced with checks which raise exceptions